devmand/gateway/src/devmand/channels/cli/QueuedCli.cpp

```diff
@@ -12,19 +12,44 @@ namespace devmand::channels::cli {
 using namespace std;
 using namespace folly;
 
+shared_ptr<QueuedCli> QueuedCli::make(
+    string id,
+    shared_ptr<Cli> cli,
+    shared_ptr<Executor> parentExecutor) {
+  return shared_ptr<QueuedCli>(new QueuedCli(id, cli, parentExecutor));
+}
+
 QueuedCli::QueuedCli(
     string _id,
     shared_ptr<Cli> _cli,
-    const shared_ptr<Executor>& _parentExecutor)
+    shared_ptr<Executor> _parentExecutor)
     : id(_id),
       cli(_cli),
+      parentExecutor(_parentExecutor),
       serialExecutorKeepAlive(SerialExecutor::create(
-          Executor::getKeepAliveToken(_parentExecutor.get()))) {}
+          Executor::getKeepAliveToken(_parentExecutor.get()))) {
+  isProcessing = false;
+  shutdown = false;
+}
 
 QueuedCli::~QueuedCli() {
-  // Execute a dummy task to make sure all running tasks have finished and that
-  // we can safely close this instance
+  MLOG(MDEBUG) << "[" << id << "] "
+               << "~QCli";
+  shutdown = true;
+  MLOG(MDEBUG) << "[" << id << "] "
+               << "~QCli: dequeuing " << queue.size() << " items";
+  QueueEntry queueEntry;
+  while (queue.try_dequeue(queueEntry)) {
+    queueEntry.promise->setException(runtime_error("QCli: Shutting down"));
+  }
+
   via(serialExecutorKeepAlive, []() {}).get();
```

Comment:
Is this still necessary ? If the lambda below captures shared_ptr to prevent destructor call before it completes ?

Reply:
We probably don't want to block until all tasks are executed - parent requesting those might be already dealocated. Throwing here should be acceptable.